lib/attrsets.nix

```
@@ -1628,26 +1628,29 @@ rec {
       binaryMerge 0 (length list);
 
   /**
-    Does the same as the update operator '//' except that attributes are
-    merged until the given predicate is verified.  The predicate should
-    accept 3 arguments which are the path to reach the attribute, a part of
-    the first attribute set and a part of the second attribute set.  When
-    the predicate is satisfied, the value of the first attribute set is
-    replaced by the value of the second attribute set.
+    Recursively updates attributes from `lhs` with attributes from `rhs`.
+
+    Returns a new attribute set containing all attributes from `lhs` and `rhs` recursively
+    In case of conflicts the value from `rhs` takes precedence over the value from `lhs`.
+
+    If the given predicate (`pred`) returns `true`, the recursion stops
+    and the value from `rhs` replaces the value from `lhs`.
 
     # Inputs
 
     `pred`
 
-    : Predicate, taking the path to the current attribute as a list of strings for attribute names, and the two values at that path from the original arguments.
+    : Predicate function (`path -> lhs -> rhs -> Bool`)
+
+      taking the path to the current attribute as a list of strings for attribute names, and the two values at that path from the original arguments.
```

While trying to formalize the nice property at the bottom of my suggestion, I realized that `lhs` was ambiguous.
Also the type signature was conflating types and identifiers.

```suggestion
    : Predicate function (of type `List String -> Any -> Any -> Bool`)

      Inputs:
      - `path : List String`: the path to the current attribute as a list of strings for attribute names
      - `lhsAtPath : Any`: the value at that path in `lhs`; same as `getAttrFromPath path lhs`
      - `rhsAtPath : Any`: the value at that path in `rhs`; same as `getAttrFromPath path rhs`
      
      Output:
      - `true`: `path` points to a value in the mapping, and `rhsAtPath` will appear in the return value of `recursiveUpdateUntil`
      - `false`: `path` is part of the nesting structure and will be an attrset in the return value of `recursiveUpdateUntil`

      `pred` is only called for `path`s that extend prefixes for which `pred` returned `false`. More formally, for any `subpath` the following holds:

      ```nix
      lib.lists.hasPrefix subpath path && subpath != path ->
        !pred subpath (getAttrFromPath subpath lhs) (getAttrFromPath subpath rhs)
      ```
```

---

Or maybe the `getAttrFromPath` equalities and the last property can be listed together as given preconditions for `pred`.
Keeps the formalized descriptions separate from the prose, might be easier to read?